features/steps/thens/values.py

```diff
@@ -1,86 +1,102 @@
 import csv
-import errors as err
 import ifcopenshell
 import os
 
-from behave import *
+from behave import register_type
 from pathlib import Path
-from utils import ifc, misc
 
+from validation_handling import gherkin_ifc, StepResult, handle_errors
 
-@then("The value must {constraint}")
-@then("The values must {constraint}")
-@then('At least "{num:d}" value must {constraint}')
-@then('At least "{num:d}" values must {constraint}')
-@err.handle_errors
-def step_impl(context, constraint, num=None):
-    errors = []
+from parse_type import TypeBuilder
+register_type(unique_or_identical=TypeBuilder.make_enum(dict(map(lambda x: (x, x), ("be unique", "be identical"))))) # todo @gh remove 'be' from enum values
+register_type(value_or_type=TypeBuilder.make_enum(dict(map(lambda x: (x, x), ("value", "type"))))) # todo @gh remove 'be' from enum values
+
+
+@gherkin_ifc.step("The value must be in '{csv_file}.csv'")
+@gherkin_ifc.step("The values must be in '{csv_file}.csv'")
+def step_impl(context, inst, csv_file):
+    if not inst:
+        return []
+
+    dirname = os.path.dirname(__file__)
+    filename = Path(dirname).parent.parent / "resources" / f"{csv_file}.csv"
+    valid_values = [row[0] for row in csv.reader(open(filename))]
+    invalid_values = [value for value in inst if value not in valid_values]
+    for value in invalid_values:
+        yield StepResult(expected = f"Value in {csv_file}.csv", observed = value)
+
+    return []
+
+
+@gherkin_ifc.step('At least "{num:d}" value must {constraint}')
+@gherkin_ifc.step('At least "{num:d}" values must {constraint}')
+def step_impl(context, inst, constraint, num):
+    stack_tree = list(
+        filter(None, list(map(lambda layer: layer.get('instances'), context._stack))))
```

Comment:
I'm a bit worried about this. We don't actually use the value of `inst` here. Wouldn't that mean that when we have multiple instances in scope at the time of execution that we report the value multiple times. Because in the decorator we already loop over the instances, but in the implementation function we get the full stack_frame.

Reply:
Maybe we should figure out a way to provide this kind of stack context from the decorator to the implementation function. We kind of set the goal for ourselves that the implementation functions would be "free" functions without the stack context.

Reply:
I've revisited the values.py. It would be indeed nice to move the stack_frame outside the step_impl. For tuples (e.g. after attribute stacking and `Its attribute Items`, such as in Scott's rule) the functionality of the stack_frame to get the right instance to inject into the db is already moved to the decorator. 
What remains is 'model-wide' functionalities. Right now, we iterate over one instance at a time. But how then do you assess that a file can only contain two IfcAlignments (silly example)? Maybe it would be an option to build in a 'model-wide' toggle to have access to all the relevant instances in a model at the time of step_impl. But I'm not sure. 
Think we can also do in in the same effort as with the other functions in values.py, see also 
https://github.com/buildingSMART/ifc-gherkin-rules/pull/125#discussion_r1444426949

Reply:
Also, the rule to which this function applies is deactivated. I'm not sure why (do you remember?), but I take it as a hint to not give it the highest priority.
https://github.com/buildingSMART/ifc-gherkin-rules/blob/edd4ffa409190c8c5f2d76a468400792b535ad6f/features/SYS001_Cable-signal-flow.feature#L22-L24

https://github.com/buildingSMART/ifc-gherkin-rules/blob/edd4ffa409190c8c5f2d76a468400792b535ad6f/features/SYS001_Cable-signal-flow.feature#L1-L5

Reply:
Yes, this was work for IfcRail that was just a proposal basically